ADR about the usage of the TLS registry in client extensions establishing TLS connection (#41126)
quarkusio/quarkus: adr/0004-using-the-tls-registry-for-clients.adoc

= Using the TLS registry for clients

* Status: proposed
* Date: 2024-06-11 by @cescoffier, @geoand, @gsmet
// * Revised:

== Context and Problem Statement

With https://github.com/quarkusio/quarkus/pull/39825[#39825], we introduced the concept of a TLS registry in Quarkus.
The TLS registry is a way to configure TLS settings in a central place and reuse them across the application and extensions.
This is particularly useful when an application needs to connect to multiple services using TLS.

Until now, each extension has _maintained_ its own way to configure TLS settings. This leads to a lot of duplication and inconsistencies.
It's also difficult for the user to understand how to configure TLS settings in Quarkus.
The TLS registry is a way to solve this problem.

The goal of this ADR is to explain how the registry should be used by client establishing a connection to a server using TLS.

> It should be noted that TLS is a sensible and complex topic.

Reviewer: ```suggestion
It should be noted that TLS is a sensitive and complex topic.
```

The TLS registry is not a silver bullet and does not solve all the problems related to TLS.
It is a way to simplify the configuration of TLS settings in Quarkus applications.

== The TLS registry

The TLS registry is an extension processing the `quarkus.tls` configuration and proposing a runtime API to access it.
Note that the configuration and the runtime API are different.
The configuration is used to define the TLS settings, while the runtime API is used to access these pre-processed settings.

> Under `quarkus.tls` can be configured the default configuration, but also _named_ configurations (`quarkus.tls.my-name`).

Reviewer: ```suggestion
When using the options directly under `quarkus.tls`, one configures the default (unamed) configuration, while using options under `quarkus.tls.my-name` configures a _named_ configurations.
```

More details can be found in the https://quarkus.io/version/main/guides/tls-registry-reference[documentation] and in the https://github.com/quarkusio/quarkus/blob/main/extensions/tls-registry/runtime/src/main/java/io/quarkus/tls/runtime/config/TlsBucketConfig.java[code].

At runtime, the https://github.com/quarkusio/quarkus/blob/main/extensions/tls-registry/runtime/src/main/java/io/quarkus/tls/TlsConfigurationRegistry.java[TlsConfigurationRegistry] CDI bean gives access to the default and named _TlsConfiguration_.
These configurations have been verified during the application startup (password, alias, validity...) and are ready to be used.
https://github.com/quarkusio/quarkus/blob/main/extensions/tls-registry/runtime/src/main/java/io/quarkus/tls/TlsConfiguration.java[TLSConfiguration] gives access to the key store, trust store, Vert.x options, tailored `SSLContext`...
It also supports reloading the certificates.

== Configuring clients with the TLS registry

When configuring a _client_ establishing a connection to a server using TLS, the TLS registry should be used to configure the TLS settings.

> The TLS settings must be defined in the `quarkus.tls.<name>` configuration.

Reviewer: ```suggestion
The TLS settings must be defined in the `quarkus.tls.<name>.` configuration.
```

                    .done();
        }
        return null;
}
----

=== Finding the named TLS configuration

Once the TLS registry is properly passed, the client extension can use it to retrieve the named TLS configuration:

[source, java]
----
TlsConfiguration configuration = null;
if (config.tlsConfigurationName.isPresent()) {
    Optional<TlsConfiguration> maybeConfiguration = tlsRegistry.get(config.tlsConfigurationName.get());
    if (!maybeConfiguration.isPresent()) {
        // We should reject missing named configuration
        throw new IllegalStateException("Unable to find the TLS configuration "
                + config.tlsConfigurationName.get() + " for the <enter client name> " + name + ".");
    }
    configuration = maybeConfiguration.get();
} else if (tlsRegistry.getDefault().isPresent() && tlsRegistry.getDefault().get().isTrustAll()) {
    defaultTrustAll = tlsRegistry.getDefault().get().isTrustAll();
    if (defaultTrustAll) {
        LOGGER.warn("The default TLS configuration is set to trust all certificates. This is a security risk."
                + "Please use a named TLS configuration for the mailer " + name + " to avoid this warning.");
    }

> }

Reviewer: We have `TlsConfiguration#from` that does the work of obtaining the proper `TlsConfiguration`

Author: Fixed.

----

Also, using the _default_ `trust-all` should be avoided, and a warning should be logged if it's used.

=== Configuring a Vert.x client

When configuring a Vert.x client (HTTP client, SMTP client, gRPC client, etc.), the `TlsConfiguration` should be used to configure the client:

> - The `keyStore` and `trustStore` should be set on the `options.setTrustOptions(config.getTrustStoreOptions())` and `options.setKeyCertOptions(config.getKeyStoreOptions())`.

Reviewer: I think it's worth mentioning that `keyStore` for clients is used for `mTLS`

}
----

- For protocol supporting SNI, the SNI should be configured using `options.setSNI(tlsConfiguration.usesSNI())`.

=== Configuring non-Vert.x clients

When configuring a non-Vert.x client, the `TLSConfiguration` provides:

- `KeyStore` objects for the key store and trust store
- A `SSLContext` object to configure the SSL context.

== Considered options

=== Continuing with the current approach

The current approach is to let each extension manage its TLS settings.
This approach leads to a lot of duplication and inconsistencies.
It's also difficult for the user to understand how to configure TLS settings in Quarkus.

Note that the TLS registry does **NOT** replace the extension-specific configuration.
It's a way to simplify the configuration of TLS settings in Quarkus applications.
The extension-specific configuration should be considered _deprecated_ and users should be invited to use the TLS registry.

== Consequences

=== Positive

* A centralized way to configure TLS settings in Quarkus applications.
* A homogenization of the TLS configuration across Quarkus extensions.
* A more complete and consistent TLS configuration.

=== Negative

> * Dealing with 2 different ways to configure TLS settings in Quarkus applications until the extension-specific configurations are removed.two

Reviewer: ```suggestion
* Dealing with 2 different ways to configure TLS settings in Quarkus applications until the extension-specific configurations are removed.
```
